runtime/lua/vim/diagnostic.lua

```diff
@@ -1566,11 +1566,14 @@ M.handlers.underline = {
         end
       end
 
+      local lines =
+        vim.api.nvim_buf_get_lines(diagnostic.bufnr, diagnostic.lnum, diagnostic.lnum + 1, true)
+
```

Comment:
Not really sure if this destroys performance. Should this be cached instead similar to `buf_line_count` metatable in `get_diagnostics`?

Reply:
instead of this, would it help to allow callers of `vim.hl.range()` to control  `eol` here? https://github.com/neovim/neovim/blob/322a6d305d088420b23071c227af07b7c1beb41a/runtime/lua/vim/hl.lua#L94-L97

Reply:
I mean would `eol=false` solve the problem. It looks like your workaround here is essentially doing that in a different way.

Reply:
should those images be swapped? your problem statement says:

> visually underline is shown starting from the next line

Reply:
Diagnostic here is set on a line with `foo = 'bar'`, right after the last `'`.

First screenshot shows that without this change there's currently no visible underline anywhere in the `foo = 'bar'` line. Instead it's shown starting only from the following `baz = 'quux',` line.

Second screenshot shows how with the change from this PR an underline being shifted to be right on the last `'` to be visible.

Not sure where the confusion comes from.

Reply:
the confusion is that the 2nd one just looks like a mistake and doesn't seem very important.

Reply:
Now I see it. Sorry. It wasn't obvious to me that "starting" sounds like it's visible there. I should've described the problem better.